power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -583,6 +583,39 @@ template <symmetry_tag sym> struct NodeState {
     }
 };
 
+class RankIteration {
+  public:
+    RankIteration(std::vector<IntS> iterations_per_rank, Idx rank_index)
+        : iterations_per_rank_{std::move(iterations_per_rank)}, rank_index_{rank_index} {}
+
+    // Getters
+    constexpr auto iterations_per_rank() const { return iterations_per_rank_; }
+    constexpr auto rank_index() const { return rank_index_; }
+
+    // Setters
+    void set_rank_index(Idx rank_index) { rank_index_ = rank_index; }
+
+    bool iterate_ranks(auto const& ranked_order, auto apply_single_object_in_rank, bool adjusted) {
+        for (Idx i = 0; i < static_cast<Idx>(ranked_order.size()); ++i) {
```

Comment:
not strictly needed but looks nicer
```suggestion
        for (Idx i = 0; i < ranked_order.ssize(); ++i) {
```